README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# ggsimplestats

<!-- badges: start -->

[![R-CMD-check](https://github.com/maximilian-heeg/ggsimplestats/workflows/R-CMD-check/badge.svg)](https://github.com/maximilian-heeg/ggsimplestats/actions)
<!-- badges: end -->

The goal of ggsimplestats is to provide an easy interface to add
statistics to boxplot created with ggplot2

## Installation

You can install the development version of ggsimplestats like so:

``` r
# install.packages("devtools")
devtools::install_github("maximilian-heeg/ggsimplestats")
```

## Minimal Example

This is a minimal example of how to add the statistics to a plot. In
this example we use the `PlantGrowth` dataset.

``` r
library(ggplot2)
library(ggsimplestats)

theme_set(ggthemes::theme_few())

ggplot(PlantGrowth, aes(group, weight, fill=group)) +
  geom_boxplot()  +
  stat_kwAllPairsDunnTest()
```

![](man/figures/README-example-1.png)<!-- -->

## Test types

Currently, four different tests are supported.

-   Two parametric tests
    -   `stat_tukeyHSD` performes a TukeyHSD posthoc test on a anova.
    -   `stat_mixedModel` creates a mixed model with `lmer` using
        `y ~ x + (1|group)`.  
        In a second step, significanes are calcuated using
        `pairs(emmeans(model ~ x))`.
-   Two non-parametric tests
    -   `stat_kwAllPairsDunnTest` uses `PMCMRplus::kwAllPairsDunnTest()`
        to perform Dunn’s non-parametric all-pairs comparison test for
        Kruskal-type ranked data
    -   `stat_frdAllPairsNemenyiTest` uses
        `PMCMRplus::frdAllPairsNemenyiTest()` to perform Nemenyi’s
        all-pairs comparisons tests of Friedman-type ranked data

## Further reading

For further information please see the `vignette("ggsimplestats")`.


NEWS.md
# ggsimplestats 0.0.1
* fixed Bug in `stat_mixedModel()` that lead to unwanted `NA` values.


# ggsimplestats 0.0.0.9000
* added `stat_tukeyHSD()`
* added `stat_kwAllPairsDunnTest()`


tests/testthat.R
library(testthat)
library(ggsimplestats)

test_check("ggsimplestats")


tests/testthat/test-stat_tukeyHSD.R
test_that("stat_tukeyHSD requires a discrete x axis", {
  library(ggplot2)

  expect_warning({
    print(ggplot(mpg, aes(cty, cty)) +
      stat_tukeyHSD())
  })
})



test_that("stat_tukeyHSD works", {
  library(ggplot2)

  data <- data.frame(
    x = c("a", "a", "a", "b", "b", "c", "c", "c", "c"),
    y = c(1, 2, 3, 4, 5, 6, 7, 8, 9)
  )

  # I expect a warning since there are ties in the dataset
  plot <- ggplot(data, aes(x, y)) +
    geom_boxplot() +
    stat_tukeyHSD()

  vdiffr::expect_doppelganger("TukeyHSD", {
    plot
  })
})


test_that("stat_tukeyHSD facet works", {
  library(ggplot2)

  plot <- ggplot(na.omit(palmerpenguins::penguins), aes(species, bill_length_mm, fill = species)) +
    geom_boxplot() +
    stat_tukeyHSD(vjust = -.2) +
    facet_grid(~sex)

  vdiffr::expect_doppelganger("TukeyHSD facet", {
    plot
  })
})

test_that("stat_tukeyHSD colors works", {
  library(ggplot2)

  data <- data.frame(
    x = c("a", "a", "a", "b", "b", "c", "c", "c", "c"),
    y = c(1, 2, 3, 4, 5, 6, 7, 8, 9)
  )

  # set seed for jitter
  set.seed(42)
  plot <- ggplot(data, aes(x, y, color=x)) +
    geom_jitter() +
    stat_tukeyHSD(colour = 'darkgreen')

  vdiffr::expect_doppelganger("stat_tukeyHSD_color", {
    plot
  })
})


tests/testthat/test-stat_mixedModel.R
test_that("stat_mixedModel works", {
  library(ggplot2)

  df <- data.frame(patient=as.factor(rep(1:5, each=4)),
                   drug=as.factor(rep(1:4, times=5)),
                   response=c(30, 28, 16, 34,
                              14, 18, 10, 22,
                              24, 20, 18, 30,
                              38, 34, 20, 44,
                              26, 28, 14, 30))
  plot <- ggplot(df, aes(x=drug, y=response, color=patient)) +
    geom_point() +
    stat_mixedModel(aes(group = patient))

  vdiffr::expect_doppelganger("stat_mixedModel_Test", {
    plot
  })
})


R/utils.R


#' Helper function to formats the results of PMCMRPlus
#'
#'
#' @param res Result from a test from formatPMCMRPlusResults
#' @param y the y position of the bracket
#'
#' @return A tibble
#' @importFrom rlang .data
#' @noRd
formatPMCMRPlusResults <- function(res, y) {
  res <- res$p.value %>%
    as.table() %>%
    as.data.frame() %>%
    # rename colums for use in ggplot
    dplyr::select(
      x = .data$Var1,
      xend = .data$Var2,
      p = .data$Freq
    ) %>%
    dplyr::mutate(
      y = y,
      x = as.numeric(as.character(.data$x)),
      xend = as.numeric(as.character(.data$xend))
    ) %>%
    dplyr::filter(!is.na(.data$p))
  return(res)
}


#' Helper function to formats the the pvalues in geomStat
#'
#'
#' @param data data.frame
#' @param step.increase Increse betweeen the different brackets
#' @param hide.ns Remove insignificant values
#'
#' @return data.frame
#' @noRd
formatPValues <- function(data, step.increase = 0.05, hide.ns = TRUE) {
  if (hide.ns) {
    data <- data[data$p < 0.05, ]
  }

  data <- data %>%
    dplyr::group_by(.data$PANEL) %>%
    ## create a y max that is a little higher than needed to expand the scales
    dplyr::mutate(ymax = .data$y + .data$range * step.increase * (dplyr::row_number()+1))

  return(data)
}


#' Helper function to adjust the y position
#'
#'
#' @param data data.frame
#' @param step.increase Increse betweeen the different brackets
#'
#' @return data.frame
#' @noRd
adjustYPosition <- function(data, step.increase) {

  data <- data %>%
    dplyr::group_by(.data$PANEL) %>%
    dplyr::mutate(y = .data$y + step.increase * dplyr::row_number())

  return(data)
}


R/geom_stat.R
#' GeomStat
#'
#' Geom layer for plotting statistics
#'
#' @import ggplot2
#' @noRd
GeomStat <- ggproto(
  "GeomStat",
  Geom,
  required_aes = c("x", "xend", "p", "y"),
  default_aes = aes(
    xend = xend
  ),
  setup_data = function(data, params) {
    data <- formatPValues(data, hide.ns = params$hide.ns, step.increase = params$step.increase)
    data
  },
  draw_panel = function(data, panel_params, coord, size, hide.ns, tick.length, format.fun, vjust, step.increase, colour) {

    # the actial y position adation takes place here
    data <- adjustYPosition(data, step.increase = (panel_params$y$limits[2]-panel_params$y$limits[1]) * step.increase )
    coords <-
      coord$transform(data, panel_params)

    # calc the lower end of the tick
    coords$ymin <- coords$y - tick.length

    # create the label for the p value
    coords$label <- format.fun(coords$p)

    label <- grid::textGrob(
      label = coords$label,
      vjust = vjust,
      x = rowMeans(cbind(coords$x, coords$xend)),
      y = coords$y,
      gp = grid::gpar(
        col = colour,
        fontsize = size
      )
    )

    bracket <- data.frame(
      x = c(coords$x, coords$x, coords$xend),
      xend = c(coords$x, coords$xend, coords$xend),
      y = c(coords$ymin, coords$y, coords$y),
      yend = c(coords$y, coords$y, coords$ymin)
    )

    bracket <- grid::segmentsGrob(
      x0 = bracket$x,
      x1 = bracket$xend,
      y0 = bracket$y,
      y1 = bracket$yend,
      gp = grid::gpar(
        col = colour
      )
    )
    return(grid::gList(label, bracket))
  },
  draw_key=NULL
)


R/stat_tukeyHSD.R
#' Add results from TukeyHSD to plot
#'
#' Uses [stats::TukeyHSD()] to calculate p values for each panel. Only works if the x-axis is a discrete scale.
#'
#' @inheritParams ggplot2::layer
#' @param geom the geometric object to use display the data. Set to `GeomStat` and usually no need to change.
#' @param size Fontsize for the annotation
#' @param na.rm If `FALSE`, the default, missing values are removed with a warning. If `TRUE`, missing values are silently removed.
#' @param hide.ns Should p-values lower than 0.05 be removed? Default TRUE
#' @param tick.length Length of the ticks in the p-value brackets. Default 0.02
#' @param format.fun A function used to format the p value. Default `scales::pvalue`
#' @param step.increase Amount of increase between two brackets. Default 0.05
#' @param vjust A numeric vector specifying vertical justification. Passed on to textGrob.
#' @param colour Colour of the bracket and label
#' @param ... Further arguments passed on to the layer in params
#'
#' @return A ggplot layer
#' @export
#'
#' @import ggplot2
#' @importFrom scales pvalue
#'
#' @examples
#' library(ggplot2)
#' library(ggsimplestats)
#'
#' theme_set(ggthemes::theme_few())
#'
#' ggplot(PlantGrowth, aes(group, weight, fill = group)) +
#'   geom_boxplot() +
#'   stat_tukeyHSD()
stat_tukeyHSD <- function(mapping = NULL, data = NULL, geom = GeomStat,
                          position = "identity", na.rm = FALSE, show.legend = NA,
                          size = 10, hide.ns = TRUE, tick.length = 0.02,
                          format.fun = pvalue, vjust = 0,
                          step.increase = 0.05, colour='black',
                          inherit.aes = TRUE, ...) {
  layer(
    stat = StatTukeyHSD, data = data, mapping = mapping, geom = geom,
    position = position, show.legend = show.legend, inherit.aes = inherit.aes,
    params = list(
      size = size, hide.ns = hide.ns, tick.length = tick.length,
      vjust = vjust, step.increase = step.increase, colour = colour,
      format.fun = format.fun, ...
    )
  )
}


#' StatTukeyHSD
#'
#' Stat layer for calculating statistics
#' Uses `TukeyHSD` from `stats`
#'
#' @import ggplot2
#' @import grid
#' @import stats
#' @import tibble
#' @importFrom tidyr separate
#' @noRd
StatTukeyHSD <- ggproto(
  "StatTukeyHSD",
  Stat,
  compute_panel = function(data, scales) {
    if (!requireNamespace("stats", quietly = TRUE)) {
      stop(
        "Package \"stats\" must be installed to use this function."
      )
    }
    if (!scales$x$is_discrete()) {
      rlang::abort(
        "x-Axis needs to be discrete."
      )
    }

    data$x <- as.factor(data$x)
    a <- aov(y ~ x, data = data)
    res <- TukeyHSD(a, "x")$x
    res <- as.data.frame(res) %>%
      tibble::rownames_to_column() %>%
      tidyr::separate(rowname, into = c("x", "xend"), sep = "-") %>%
      dplyr::select(x, xend, p = `p adj`)


    res$x <- as.numeric(as.character(res$x))
    res$xend <- as.numeric(as.character(res$xend))
    res$y <- max(data$y)
    res$range <- max(data$y) - min(data$y)

    return(res)
  },
  required_aes = c("x", "y")
)


R/stat_kwAllPairsDunnTest.R
#' Add results from Dunn's non-parametric test to plot
#'
#' Uses [PMCMRplus::kwAllPairsDunnTest()] to perform Dunn's non-parametric
#' all-pairs comparison test for Kruskal-type ranked data and adds the
#' resulting p-values to the plot.
#' Only works if the x-axis is a discrete scale.
#'
#' @inheritParams ggplot2::layer
#' @param geom the geometric object to use display the data. Set to `GeomStat` and usually no need to change.
#' @param size Fontsize for the annotation
#' @param na.rm If `FALSE`, the default, missing values are removed with a warning. If `TRUE`, missing values are silently removed.
#' @param hide.ns Should p-values lower than 0.05 be removed? Default TRUE
#' @param tick.length Length of the ticks in the p-value brackets. Default 0.02
#' @param format.fun A function used to format the p value. Default `scales::pvalue`
#' @param step.increase Amount of increase between two brackets. Default 0.05
#' @param vjust A numeric vector specifying vertical justification. Passed on to textGrob.
#' @param colour Colour of the bracket and label
#' @param ... Further arguments passed on to the layer in params
#'
#' @return A ggplot layer
#' @export
#'
#' @import ggplot2
#' @importFrom scales pvalue
#'
#' @examples
#' library(ggplot2)
#' library(ggsimplestats)
#'
#' theme_set(ggthemes::theme_few())
#'
#' ggplot(PlantGrowth, aes(group, weight, fill = group)) +
#'   geom_boxplot() +
#'   stat_kwAllPairsDunnTest()
stat_kwAllPairsDunnTest <- function(mapping = NULL, data = NULL, geom = GeomStat,
                                    position = "identity", na.rm = FALSE, show.legend = NA,
                                    size = 10, hide.ns = TRUE, tick.length = 0.02,
                                    format.fun = pvalue, vjust = 0,
                                    step.increase = 0.05,colour='black',
                                    inherit.aes = TRUE, ...) {
  layer(
    stat = StatKwAllPairsDunnTest, data = data, mapping = mapping, geom = geom,
    position = position, show.legend = show.legend, inherit.aes = inherit.aes,
    params = list(
      size = size, hide.ns = hide.ns, tick.length = tick.length,
      vjust = vjust, step.increase = step.increase, colour = colour,
      format.fun = format.fun, ...
    )
  )
}


#' StatKwAllPairsDunnTest
#'
#' Stat layer for calculating statistics
#' Uses `StatKwAllPairsDunnTest` from `PMCMRplus`
#'
#' @import ggplot2
#' @import grid
#' @noRd
StatKwAllPairsDunnTest <- ggproto(
  "StatKwAllPairsDunnTest",
  Stat,
  compute_panel = function(data, scales) {
    if (!requireNamespace("PMCMRplus", quietly = TRUE)) {
      stop(
        "Package \"PMCMRplus\" must be installed to use this function."
      )
    }
    if (!scales$x$is_discrete()) {
      rlang::abort(
        "x-Axis needs to be discrete."
      )
    }



    res <- PMCMRplus::kwAllPairsDunnTest(
      x = data$y,
      g = data$x
    )
    res <- formatPMCMRPlusResults(res, y = max(data$y))
    res$range <- max(data$y) - min(data$y)

    return(res)
  },
  required_aes = c("x", "y")
)


R/stat_frdAllPairsNemenyiTest.R
#' Add results Nemenyi's All-Pairs Comparisons test to plot
#'
#' Uses [PMCMRplus::frdAllPairsNemenyiTest()] to perform Nemenyi's all-pairs
#' comparisons tests of Friedman-type ranked data and adds the
#' resulting p-values to the plot. The grouping aesthetics will be used as the
#' block for `frdAllPairsNemenyiTest`.
#' Only works if the x-axis is a discrete scale.
#'
#' @inheritParams ggplot2::layer
#' @param geom the geometric object to use display the data. Set to `GeomStat` and usually no need to change.
#' @param size Fontsize for the annotation
#' @param na.rm If `FALSE`, the default, missing values are removed with a warning. If `TRUE`, missing values are silently removed.
#' @param hide.ns Should p-values lower than 0.05 be removed? Default TRUE
#' @param tick.length Length of the ticks in the p-value brackets. Default 0.02
#' @param format.fun A function used to format the p value. Default `scales::pvalue`
#' @param step.increase Amount of increase between two brackets. Default 0.05
#' @param vjust A numeric vector specifying vertical justification. Passed on to textGrob.
#' @param colour Colour of the bracket and label
#' @param ... Further arguments passed on to the layer in params
#'
#' @return A ggplot layer
#' @export
#'
#' @import ggplot2
#' @importFrom scales pvalue
#'
#' @examples
#' library(ggplot2)
#' library(ggsimplestats)
#'
#' theme_set(ggthemes::theme_few())
#'
#' df <- data.frame(patient=as.factor(rep(1:5, each=4)),
#'                  drug=as.factor(rep(1:4, times=5)),
#'                  response=c(30, 28, 16, 34,
#'                             14, 18, 10, 22,
#'                             24, 20, 18, 30,
#'                             38, 34, 20, 44,
#'                             26, 28, 14, 30))
#' ggplot(df, aes(x=drug, y=response, color=patient)) +
#'   geom_point(position = position_dodge(width = .2)) +
#'   stat_frdAllPairsNemenyiTest(aes(group = patient))
stat_frdAllPairsNemenyiTest <- function(mapping = NULL, data = NULL, geom = GeomStat,
                                    position = "identity", na.rm = FALSE, show.legend = NA,
                                    size = 10, hide.ns = TRUE, tick.length = 0.02,
                                    format.fun = pvalue, vjust = 0,
                                    step.increase = 0.05,colour='black',
                                    inherit.aes = TRUE, ...) {
  layer(
    stat = StatFrdAllPairsNemenyiTest, data = data, mapping = mapping, geom = geom,
    position = position, show.legend = show.legend, inherit.aes = inherit.aes,
    params = list(
      size = size, hide.ns = hide.ns, tick.length = tick.length,
      vjust = vjust, step.increase = step.increase, colour = colour,
      format.fun = format.fun, ...
    )
  )
}


#' StatFrdAllPairsNemenyiTest
#'
#' Stat layer for calculating statistics
#' Uses `frdAllPairsNemenyiTest` from `PMCMRplus`
#'
#' @import ggplot2
#' @import grid
#' @noRd
StatFrdAllPairsNemenyiTest <- ggproto(
  "StatFrdAllPairsNemenyiTest",
  Stat,
  compute_panel = function(data, scales) {
    if (!requireNamespace("PMCMRplus", quietly = TRUE)) {
      stop(
        "Package \"PMCMRplus\" must be installed to use this function."
      )
    }
    if (!scales$x$is_discrete()) {
      rlang::abort(
        "x-Axis needs to be discrete."
      )
    }



    res <- PMCMRplus::frdAllPairsNemenyiTest(
      y = data$y,
      g = data$x,
      blocks = data$group
    )
    res <- formatPMCMRPlusResults(res, y = max(data$y))
    res$range <- max(data$y) - min(data$y)

    return(res)
  },
  required_aes = c("x", "y", "group")
)


R/stat_mixedModel.R
#' Add results linear mixed model to plot
#'
#' Uses [lme4::lmer()] and [emmeans::emmeans()] to calcuate significanes.
#' First a mixed model is build with `lmer` using `y ~ x + (1|group)`.
#' In a second step, significanes are calcuated using `pairs(emmeans(model ~ x))`.
#' Only works if the x-axis is a discrete scale.
#'
#' @inheritParams ggplot2::layer
#' @param geom the geometric object to use display the data. Set to `GeomStat` and usually no need to change.
#' @param size Fontsize for the annotation
#' @param na.rm If `FALSE`, the default, missing values are removed with a warning. If `TRUE`, missing values are silently removed.
#' @param hide.ns Should p-values lower than 0.05 be removed? Default TRUE
#' @param tick.length Length of the ticks in the p-value brackets. Default 0.02
#' @param format.fun A function used to format the p value. Default `scales::pvalue`
#' @param step.increase Amount of increase between two brackets. Default 0.05
#' @param vjust A numeric vector specifying vertical justification. Passed on to textGrob.
#' @param colour Colour of the bracket and label
#' @param ... Further arguments passed on to the layer in params
#'
#' @return A ggplot layer
#' @export
#'
#' @import ggplot2
#' @importFrom scales pvalue
#'
#' @examples
#' library(ggplot2)
#' library(ggsimplestats)
#'
#' theme_set(ggthemes::theme_few())
#'
#' df <- data.frame(patient=as.factor(rep(1:5, each=4)),
#'                  drug=as.factor(rep(1:4, times=5)),
#'                  response=c(30, 28, 16, 34,
#'                             14, 18, 10, 22,
#'                             24, 20, 18, 30,
#'                             38, 34, 20, 44,
#'                             26, 28, 14, 30))
#' ggplot(df, aes(x=drug, y=response, color=patient)) +
#'   geom_point(position = position_dodge(width = .2)) +
#'   stat_mixedModel(aes(group = patient))
stat_mixedModel <- function(mapping = NULL, data = NULL, geom = GeomStat,
                                        position = "identity", na.rm = FALSE, show.legend = NA,
                                        size = 10, hide.ns = TRUE, tick.length = 0.02,
                                        format.fun = pvalue, vjust = 0,
                                        step.increase = 0.05,colour='black',
                                        inherit.aes = TRUE, ...) {
  layer(
    stat = StatMixedModel, data = data, mapping = mapping, geom = geom,
    position = position, show.legend = show.legend, inherit.aes = inherit.aes,
    params = list(
      size = size, hide.ns = hide.ns, tick.length = tick.length,
      vjust = vjust, step.increase = step.increase, colour = colour,
      format.fun = format.fun, ...
    )
  )
}


#' StatMixedModel
#'
#' Stat layer for calculating statistics
#' Uses [lme4::lmer()] and [emmeans::emmeans()]
#'
#' @import ggplot2
#' @import grid
#' @importFrom tidyr separate
#' @noRd
StatMixedModel <- ggproto(
  "StatMixedModel",
  Stat,
  compute_panel = function(data, scales) {
    if (!requireNamespace("lme4", quietly = TRUE)) {
      stop(
        "Package \"lme4\" must be installed to use this function."
      )
    }
    if (!requireNamespace("emmeans", quietly = TRUE)) {
      stop(
        "Package \"emmeans\" must be installed to use this function."
      )
    }
    if (!requireNamespace("stringr", quietly = TRUE)) {
      stop(
        "Package \"stringr\" must be installed to use this function."
      )
    }
    if (!scales$x$is_discrete()) {
      rlang::abort(
        "x-Axis needs to be discrete."
      )
    }

    data$x <- as.factor(data$x)
    data$group <- as.factor(data$group)

    model <- lme4::lmer(y ~ x + (1|group), data = data)
    res <- graphics::pairs(emmeans::emmeans(model, ~ x)) %>%  as.data.frame()

    res <- res  %>%
      tidyr::separate(contrast, into = c("x", "xend"), sep = " - ") %>%
      dplyr::select(x, xend, p = `p.value`)

    res$x <- as.numeric(substring(as.character(res$x),2))
    res$xend <- as.numeric(substring(as.character(res$xend),2))
    res$y <- max(data$y)
    res$range <- max(data$y) - min(data$y)

    return(res)
  },
  required_aes = c("x", "y", "group")
)
